### PR TITLE
Allow building an upgrade from an async function

### PR DESCRIPTION
**Pull request by tomaka**
This is a quality-of-life change, with the objective of trying to make rust-libp2p easier to use.

I think it's quite nice and comprehensible to be able to write upgrades as shown in the doc-comment example.

### CODE REVIEW

File: core/src/upgrade/from_fn.rs [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+///         upgrade::apply(out, upgrade::from_fn("/foo/1", move |mut sock, endpoint| async move {
+///             if endpoint.is_dialer() {
+///                 upgrade::write_one(&mut sock, "some handshake data").await?;
+///             } else {
+///                 let handshake_data = upgrade::read_one(&mut sock, 1024).await?;
+///                 if handshake_data != b"some handshake data" {
+///                     return Err(upgrade::ReadOneError::from(io::Error::from(io::ErrorKind::Other)));
+///                 }
+///             }
+///
```
Comment (twittner) [inline core/src/upgrade/from_fn.rs]:
```suggestion
```

File: core/src/upgrade/from_fn.rs [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+impl<C, P, F, Fut, Err, Out> InboundUpgrade<C> for FromFnUpgrade<P, F>
+where
+    P: ProtocolName + Clone,
+    F: FnOnce(Negotiated<C>, Endpoint) -> Fut,
+    Fut: Future<Output = Result<Out, Err>>,
+{
+    type Output = Out;
+    type Error = Err;
+    type Future = Pin<Box<Fut>>; // TODO: don't box the future
```
Comment (twittner) [inline core/src/upgrade/from_fn.rs]: Do you want to file an issue for the TODOs in here?

Comment (tomaka) [reply core/src/upgrade/from_fn.rs]: We could also do #1384 first, I'm not sure.

Comment (tomaka) [reply core/src/upgrade/from_fn.rs]: Will rebase on top of #1384 once it's merged.

**Comment by tomaka** [comment]
Updated and concerns addressed!

**Comment by tomaka** [comment]
Any objection to merging?

**Comment by twittner** [comment]
Not at all.

**Pull request merged**